caproto/server/stats.py

```diff
@@ -0,0 +1,562 @@
+            await self.application_directory.write(value=str(main_path.parent))
+        except TypeError:
+            # Built-in is OK
+            ...
+        except Exception:
+            self.log.exception('Unable to determine startup directory')
+
+        if psutil is not None:
+            await self.cpu_count.write(value=psutil.cpu_count())
+
+
+class PeriodicStatusHelper(PVGroup):
+    """
+    An IocStats-like tool for caproto IOCs.  Includes values which update
+    on a periodic basis (:attr:`update_period`, PV ``UPD_TIME``).
+    """
+
+    _process: Optional['psutil.Process']
+    _root_pvgroup: PVGroup
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._root_pvgroup = _find_top_level_pvgroup(self)
+        self._process = None
+
+        if psutil is None:
+            self.log.warning(
+                "The Python library psutil is not installed, so MEM_USED will "
+                "not be reported by StatsHelper."
+            )
+        else:
+            self._process = psutil.Process()
+
+    access = pvproperty(
+        # Not implemented:
+        name='ACCESS',
+        doc='CA Security access level to this IOC',
+        enum_strings=['Running', 'Maintenance', 'Test', 'OFFLINE'],
+        record='mbbo',
+        value='Running',
+        dtype=ChannelType.ENUM,
+    )
+
+    start_tod = pvproperty(
+        value='',
+        max_length=100,
+        name='STARTTOD',
+        record='stringin',
+        doc="Time and date of startup",
+        read_only=True,
+    )
+
+    @start_tod.startup
+    async def start_tod(self, instance, async_lib):
+        await self.start_tod.write(value=str(datetime.datetime.now()))
+
+    time_of_day = pvproperty(
+        value='',
+        name='TOD',
+        max_length=100,
+        record='stringin',
+        doc="Current time and date",
+        read_only=True
+    )
+
+    @time_of_day.scan(period=1)
+    async def time_of_day(self, instance, async_lib):
+        """Update time-of-day."""
+        await self.time_of_day.write(value=str(datetime.datetime.now()))
+
+    heartbeat = pvproperty(
+        value=0,
+        name='HEARTBEAT',
+        record='calcout',
+        doc='1 Hz counter since startup',
+        read_only=True,
+    )
+
+    @heartbeat.scan(period=1)
+    async def heartbeat(self, instance, async_lib):
+        await self.heartbeat.write(value=1 - self.heartbeat.value)
+
+    start_count = pvproperty(
+        value=0,
+        name='START_CNT',
+        record='calcout',
+        doc='Startup count, if autosave is working',
+        read_only=True,
+    )
+    autosaved(start_count)
+
+    @start_count.startup
+    async def start_count(self, instance, async_lib):
+        # Give autosave some time to load
+        await async_lib.library.sleep(3)
+        await self.start_count.write(value=self.start_count.value + 1)
+
+    # TODO: any way we can find this information?
```

Comment:
This is tough. The number of connected clients is available via the logging stream, so at least there is a way to know it, but it would require significant changes to thread that into here. `run(...)` consumes the `pvdb` dict and has no way to inject itself in.

In fact, making these objects aware of their server might be abstraction-breaking, as a given `pvdb` could conceivably be run by more than one server. That's a wild and probably useless possibility, but just a way of saying there is currently no 1-to-1 coupling here and implementing this would force us to introduce it.

Reply:
Regarding abstractions, your thoughts mirror my own. I rather like that they are separated, but its effect is unfortunate in this case (and also the case of allowing classes access to `async_lib` without stashing it in a `startup` method).

Reply:
It's worth thinking about in the long term whether we can keep most of this separation but make it possible to handle these issues. One possibility is using context vars to support a `what_is_my_server()`function. I would want to sit with that for awhile though; it feels a bit too fancy.